nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ScriptedTransformRecord.java

```diff
@@ -0,0 +1,398 @@
+                    attributesToAdd.put("mime.type", writer.getMimeType());
+                    attributesToAdd.putAll(writeResult.getAttributes());
+                    attributesToAdd.put("record.count", String.valueOf(writeResult.getRecordCount()));
+                } catch (final MalformedRecordException | SchemaNotFoundException | ScriptException e) {
+                    throw new ProcessException(e);
+                }
+            });
+
+            // Add the necessary attributes to the FlowFile and transfer to success
+            session.putAllAttributes(flowFile, attributesToAdd);
+            session.transfer(flowFile, REL_SUCCESS);
+
+            final long transformCount = recordIndex.get() - dropCount.get();
+            getLogger().info("Successfully transformed {} Records and dropped {} Records for {}", new Object[] {transformCount, dropCount.get(), flowFile});
+            session.adjustCounter("Records Transformed", transformCount, true);
+            session.adjustCounter("Records Dropped", dropCount.get(), true);
+
+            final long millis = System.currentTimeMillis() - startMillis;
+            session.getProvenanceReporter().modifyContent(flowFile, "Transformed " + transformCount + " Records, Dropped " + dropCount.get() + " Records", millis);
+        } catch (final ProcessException e) {
+            getLogger().error("After processing {} Records, encountered failure when attempting to transform {}", new Object[] {recordIndex.get(), flowFile}, e.getCause());
+            session.transfer(flowFile, REL_FAILURE);
+        } catch (final Exception e) {
+            getLogger().error("After processing {} Records, encountered failure when attempting to transform {}", new Object[] {recordIndex.get(), flowFile}, e);
+            session.transfer(flowFile, REL_FAILURE);
+        }
+    }
+
+    private ScriptEvaluator createEvaluator(final ScriptEngine scriptEngine, final FlowFile flowFile) throws ScriptException {
+        if (PYTHON_SCRIPT_LANGUAGE.equalsIgnoreCase(scriptEngine.getFactory().getLanguageName())) {
+            return new PythonScriptEvaluator(scriptEngine, scriptToRun, flowFile);
+        }
+
+        return new DefaultScriptEvaluator(scriptEngine, scriptToRun, flowFile);
+    }
+
+    private static Bindings setupBindings(final ScriptEngine scriptEngine) {
+        Bindings bindings = scriptEngine.getBindings(ScriptContext.ENGINE_SCOPE);
+        if (bindings == null) {
+            bindings = new SimpleBindings();
+        }
+
+        scriptEngine.setBindings(bindings, ScriptContext.ENGINE_SCOPE);
+
+        return bindings;
+    }
+
+
+    @Override
+    public Collection<SearchResult> search(final SearchContext context) {
+        return ScriptingComponentUtils.search(context, getLogger());
+    }
+
+    private interface ScriptEvaluator {
+        Object evaluate(Record record, long index) throws ScriptException;
+    }
+
+
+    private class PythonScriptEvaluator implements ScriptEvaluator {
```

Comment:
@markap14 curious why we're not pre-compiling Groovy and other languages as well.

Reply:
The ScriptEngine API doesn't support pre-compiling the script - and shouldn't. That should be handled by the Script Engine itself. It is in the case of Groovy and I would suppose others. Unfortunately, the Jython engine doesn't bother with this, and it's up to the implementors to do it. Honestly, the fact that we cast the script engine into a particular implementation and pre-compile for Jython is a hack. But went with it anyway because a quick benchmark on my laptop showed a performance improvement of well over 10x for a simple script. And Jython is likely to be the most commonly used language, I would guess.

Reply:
We could check the instance to see if it is `Compilable` and compile it if so, but IIRC some engines don't mark themselves as Compilable but they always compile under the hood anyway.

Reply:
Ah, I did not realize that there was a `Compilable` interface. That definitely is cleaner. I can update to use that.

Reply:
@markap14 FYI, there is a bug in the version of JRuby that we're using that caused me some really weird issues when I played around with NIFI-5422. Still working on that one. Here's the issue I raised. They think they have a fix, but we'd need to test:

https://github.com/jruby/jruby/issues/6230

Reply:
Thanks @MikeThomsen . Well I gave this a try with CompiledScript. Unfortunately, it looks like JRuby has another problem in that it ignores the bindings that are passed to it. So I'll avoid going the route of CompiledScript for now, and just use it the way that I was previously. In testing this though, I found a bug where I was still constantly recompiling the Jython script when I didn't need to be. So will push an update for htat.

Reply:
I think we'll need to move toward some abstraction which does the evaluation either with a compiledscript or with invocable as is appropriate. Groovy would definitely benefit from the compiledscript option, the rest are iffy. For Kotlin it was an absolute requirement to compile the last time I played with it.

Reply:
I don't believe Groovy will benefit from compiling, because the Groovy Engine itself already handles pre-compiling it and caching a Map of Script text to CompiledScript object. I would probably shy away from creating another abstraction that performs Invocable vs. CompileScript, etc. and instead just push for the teams maintaining the Engines themselves to address the concerns.

Reply:
... and after re-reading some of the JavaDocs, perhaps it does make sense. Unfortunately, the javax.script package has a lot of these "magical interfaces." We can probably do some refactoring and abstraction in a way that can handle pre-compiling where necessary and using the most appropriate interface. Personally I think this is a weakness in the API, but it is the API that we have. So another layer of abstraction may at least allow us to address it once instead of in each processor, etc.

I do think that is beyond the scope of this Jira, though. Would make sense in a follow-on Jira.

Reply:
@markap14 Scala and Kotlin are both JSR223-capable, but would definitely need to be compiled. When I experimented locally with Kotlin and I found that the engine appeared to have very different performance numbers with Invocable vs Compilable.

The guidance from the JRuby team is that all it does is save you from reparsing the scripts, but even that would definitely have value. We already have a ticket to track some of that [here](https://issues.apache.org/jira/browse/NIFI-5422).

I agree that the changes should be made here, but they're worth exploring in other Jiras.